### ISSUE TITLE
[BUG] DummyForecaster interface is inconsistent

### ISSUE THREAD

**Issue by big-o** [alan-turing-institute/sktime#199]
**Describe the bug**

`DummyForecaster` does not adhere to the forecaster interface. The forecast horizon is given to `fit` and the predictions are calculated in this method instead of in `predict`. This prevents your from using this forecaster in the same interfaces and pipelines as everything else.

**To Reproduce**

```python
def do_forecast(ts, fcstr, test_size=20):
    train, test = ts[:-test_size], ts[-test_size:]
    fh = np.arange(len(test)) + 1
    
    fcstr.fit(train)
    fcstr.predict(fh=fh)
    
data = load_airline()

fcstrs = [
    ThetaForecaster(),
    ExpSmoothingForecaster(),
    ARIMAForecaster(),
    DummyForecaster()
]
for fcstr in fcstrs:
    # ValueError: DummyForecaster requires to specify the forecasting horizon in `fit`
    do_forecast(data, fcstr)
```

**Expected behavior**

`DummyForecaster.fit()` should not require a forecast horizon. Predictions should be done in `DummyForecaster.predict()`.

**Additional context**

**Versions**
<details>

Python 3.6.8 (default, Oct  7 2019, 12:59:55) 
[GCC 8.3.0]
NumPy 1.17.4
SciPy 1.3.2
sktime 0.3.1
</details>

**Comment by mloning** [comment]
Thanks for raising the issue! Some forecasters will need the forecasting horizon already during training (e.g. reduction to regression when you fit in training one estimator for each step of the forecasting horizon).  

In the DummyForecaster that's probably not the case once it's more cleanly implemented, it's still experimental, I needed it in fit for the "seasonal last" strategy I believe, but the different strategies need to be separated more cleanly into different classes, e.g. into a TrendForecaster which also handles seasonal trends and perhaps some kind of a SmoothingForecaster (for rolling mean predictions, etc).

**Comment by fkiraly** [comment]
Is this maybe also another interesting architecture discussion point raised?
- interaction of different kinds of forecasters and forecasting, with inheritance hierarchy and task specification and/or method signature

**Comment by big-o** [comment]
I see your point, but looking at the code it seems to me like `fh` is only really being used to calculate `y_pred` - something that I think could (and should) be put off until the call to `predict`. The only real "fitting" I can see in any of the strategies is in the "linear" strategy, where a SARIMAX estimator is trained, but this estimator makes no use of `fh` except, again, when calculating the actual predictions.

Unless i've misunderstood something, if you delay calculating `y_pred` until you're inside the `predict` method (which is the behaviour I would have expected), I don't think there's any need for `fh` in _any_ of the current strategies until `predict` is called, as per the other forecasters.

**Comment by fkiraly** [comment]
The most common example of a strategy that needs `fh` already within fit is the lagged/rolling window reduction to supervised learning. `fh` is needed within fit to create the vector/matrix of the y-s to present the tabular supervised learner with.

**Comment by big-o** [comment]
Just to clarify for everyone, the point I want to make here is that I don't think `fh` is needed by `DummyForecaster.fit` specifically - I won't comment on what might be needed by other/future forecasters. I realise now that the way I worded it might have made it sound like I don't think `fh` should be needed by _any_ forecaster `fit()` method, but as you've already pointed out this isn't the case.

**Comment by fkiraly** [comment]
Ah, I see - though still, the "unified interface" architectural principle usually implies that all methods belonging to a class of strategies should support methods with the same (key) signature, as to not break if you iterate over methods or switch them out.

E.g., in sklearn, k-nearest-neighors still has a `fit` methods, even though all it does is to remember the data for `predict`.

Of course the jury is still out on where `fh` should sit - in the native fit, or in a task object, somehwere else etc.

**Comment by big-o** [comment]
In lieu of joining your discussion later this week, I'll leave a few thoughts here about this issue.

I think the problem here is inconsistent method signatures across `BaseForecaster` children, which breaks the duck-typing approach that many people like to use: if it implements `fit` and `predict`, then I should be able to use it in the same way as anything else that implements `fit` and `predict`. This inconsistency also makes much it harder to use pipelines and `Mixin`s effectively.

Having thought about it more, I now think that any extra parameters that are needed at fit-time should be provided to the constructor instead, meaning that `fit` never takes anything except the training data, regardless of the algorithm used. So things like `DummyForecaster`, which currently requires knowledge of the horizon at fit-time, would now be given `fh` as a constructor parameter instead. If you need to train the `DummyForecaster` with a different horizon, then you would have to create a new instance (which I think is clearer and encourages less error-prone coding by the user anyway).

Likewise for predict: it should ideally take nothing more than a forecast horizon, optional exogenous variables and confidence levels for the desired PIs (since we don't want to have to retrain a new model just to change the PIs). In the case of the `DummyForecaster`, since we provided `fh` on initialisation, all `predict` would do here is validate that the provided `fh` is the same as the initial one. Other forecasters can be more flexible as no prior constraints have been placed on `fh`. But despite this difference, all forecasters still maintain the same _interface_, so I can use them consistently.

This consistency will also make it easier to take more of a `Mixin` approach: take a look at the RegressorMixin in sklearn as a good example: [it assumes `predict()` is implemented in a certain way](https://github.com/scikit-learn/scikit-learn/blob/1495f6924/sklearn/base.py#L408).

**Comment by fkiraly** [comment]
Interesting.
I'll comment separately on multiple points raised.

(i) "fit/predict should have a unified/consistent interface" - I completely agree with this, though I would restrict this requirement within "task types", of which there are multiple.
For example, a supervised learner will naturally expects a different input (X/y data frames) than a forecaster (series). But of course within the "task type", all learners should adhere to the same specification standard for methods, interfaces, etc.

(ii) "stick `fh` in the constructor" - my intuition says "don't" here, since there are methods that don't need a `fh` specified that early.

Now I get that your recommendation for (ii) comes from (i) - but there is another way to comply with (i), namely having a "control" structured container, possibly but not necessarily a "task object". Another alternative is polymorphic `fit`/`predict` signatures with dispatch methods.

We have indeed been discussing mixins, in the context of these last two - mixins for "task type" specific dispatch.

(iii) "specify PI confidence in `predict`" - the rationale for this recommendation of yours seems to clash with that for `fh` - why: there are models which need PI already specified in `fit`. For example: models fit using a Q-loss aka pinball loss for that specific confidence level. Thus, if your argumentation were consistent, you would want the PI to be specified at construction (which I would somewhat disagree with in line with my comment about (ii) above).

**Comment by big-o** [comment]
> (i) "fit/predict should have a unified/consistent interface" - I completely agree with this, though I would restrict this requirement within "task types", of which there are multiple.
> For example, a supervised learner will naturally expects a different input (X/y data frames) than a forecaster (series). But of course within the "task type", all learners should adhere to the same specification standard for methods, interfaces, etc.

Absolutely - I think this is more or less already the case, unless I'm mistaken? Just there are a few exceptions to the rule that aren't being adhered to (like the DummyForecaster).

> (ii) "stick fh in the constructor" - my intuition says "don't" here, since there are methods that don't need a fh specified that early.

Again this suggestion was specific to DummyForecaster (or any other forecaster that needs `fh` before `predict`). The only thing I'd want to enforce in this approach is specifiying `fh` in `predict` and not in `fit`. `__init__` would be more relaxed and allow you to code in the exceptions to the rule, such as forecasters that need to know `fh` beforehand. So in the case of DummyForecaster you'd be specifying `fh` twice, which seems a little wasteful but it's there to guarantee compliance with the interface. Other forecasters would only need to specify `fh` in the `predict` method.

**Comment by fkiraly** [comment]
Valid point. Just to play the devil's advocate: can anyone think of a situation where it is of substantial advantage to specify `fh` late in `predict` and not already in `fit`?

It's true that some learners don't need it, but is there (beyond usabiluty) any advantage to not supply it, e.g., in deployment scenarios?

**Comment by mloning** [comment]
Fixed in #218
